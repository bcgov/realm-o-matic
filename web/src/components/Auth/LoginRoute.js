// eslint-disable-next-line
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import implicitAuthManager from '../../utils/auth';

export const LoginRoute = ({ match }) => {
  const idp = match.params.idp;
  if (idp) implicitAuthManager.config.kcIDPHint = idp;
  window.location.assign(implicitAuthManager.getSSOLoginURI());
  return <div>Your are being redirected to KeyCloak login page...</div>;
};

LoginRoute.propTypes = {
  match: PropTypes.object.isRequired,
};
