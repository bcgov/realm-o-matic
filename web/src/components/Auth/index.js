export { default as AuthModal } from './AuthModal';
export { LoginRoute } from './LoginRoute';
export { default as ProtectedRoute } from './ProtectedRoute';
