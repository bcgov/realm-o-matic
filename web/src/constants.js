//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

import config from './config.json';

export const SSO_CONFIG = {
  baseURL: process.env.REACT_APP_SSO_BASE_URL || config.ssoBaseUrl,
  realmName: process.env.REACT_APP_SSO_REALM_NAME || config.ssoRealmName,
  clientId: process.env.REACT_APP_SSO_CLIENT_ID || config.ssoClientId,
};

export const API = {
  BASE_URL: () =>
    process.env.NODE_ENV === 'development'
      ? 'http://localhost:8000/api/v1/'
      : `${window.location.origin}/api/v1/`,
  TIME_OUT: 40000,
  IDP: () => 'idps',
};

export const TEST_IDS = {
  APP: {
    LOGIG: 'login-button',
    GET_IDPS: 'get-idps-button',
  },
};

export const APP_INFO = {
  NAME: 'Realm-o-Matic',
};

export const SSO_IDP = {
  GITHUB: 'github',
  IDIR: 'idir',
};

export const FOOTER_LINKS = [
  {
    name: 'Home',
    link: '.',
  },
  {
    name: 'Disclaimer',
    link: 'https://www2.gov.bc.ca/gov/content/home/disclaimer',
  },
  {
    name: 'Privacy',
    link: 'https://www2.gov.bc.ca/gov/content/home/privacy',
  },
  {
    name: 'Accessibility',
    link: 'https://www2.gov.bc.ca/gov/content/home/accessibility',
  },
  {
    name: 'Copyright',
    link: 'https://www2.gov.bc.ca/gov/content/home/copyright',
  },
  {
    name: 'Contact Us',
    link: 'https://github.com/bcgov/realm-o-matic',
  },
];
