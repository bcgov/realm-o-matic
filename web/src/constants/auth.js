//
// Realm-o-Matic
//
// Copyright Â© 2019 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Shelly Xue Han
//

import config from '../config.json';

export const SSO_CONFIG = {
  baseURL: process.env.REACT_APP_SSO_BASE_URL || config.ssoBaseUrl,
  realmName: process.env.REACT_APP_SSO_REALM_NAME || config.ssoRealmName,
  clientId: process.env.REACT_APP_SSO_CLIENT_ID || config.ssoClientId,
};

export const SSO_IDP = {
  GITHUB: 'github',
  IDIR: 'idir',
};

export const ACCESS_CONTROL = {
  REQUESTER_ROLE: 'devhub_kc_requester',
  REVIEWER_ROLE: 'devhub_kc_reviewer',
  NO_ROLE: 'unauthorized',
};
