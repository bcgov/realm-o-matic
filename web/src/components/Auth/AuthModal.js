import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { Link } from 'react-router-dom';
import { Modal, Button } from 'semantic-ui-react';
import { SSO_IDP } from '../../constants/auth';
import { TEST_IDS } from '../../constants/ui';

const StyledModal = styled(Modal)`
  &&& {
    width: 400px;
    text-align: center;
    padding: 10px 0;
  }
  .content {
    margin: 30px 0;
  }
  .ui.button {
    color: white;
    background-color: #036;
  }
`;

export const AuthModal = ({ isAuthenticated }) => {
  const buttons = (
    <Button.Group onActionClick={null}>
      <Link to={{ pathname: `/login/${SSO_IDP.GITHUB}` }}>
        <Button data-testid={TEST_IDS.AUTH.GITHUB_LOGIN}>GitHub</Button>
      </Link>
      <Button.Or />
      <Link to={{ pathname: `/login/${SSO_IDP.IDIR}` }}>
        <Button data-testid={TEST_IDS.AUTH.IDIR_LOGIN}>IDIR</Button>
      </Link>
    </Button.Group>
  );

  const options = {
    open: !isAuthenticated,
    closeOnEscape: false,
    closeOnDimmerClick: false,
    size: 'tiny',
    dimmer: 'blurring',
    header: 'Welcome to Realm-o-Matic',
    content: 'Please login with one of the following options before proceeding to the app.',
    actions: buttons,
  };
  return <StyledModal data-testid={TEST_IDS.AUTH.MODAL} {...options} />;
};

AuthModal.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
};

export default AuthModal;
