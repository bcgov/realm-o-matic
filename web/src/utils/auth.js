import { ImplicitAuthManager } from '@bcgov/common-web-utils';
import { SSO_CONFIG } from '../constants/auth';

const iam = new ImplicitAuthManager(SSO_CONFIG);

export default iam;
