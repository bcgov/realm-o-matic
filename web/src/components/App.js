import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Route, Switch } from 'react-router-dom';
import implicitAuthManager from '../utils/auth';
import { authenticateFailed, authenticateSuccess } from '../actions';
import { Auth, Home, Restricted, NewRequest } from '../containers';
import { LoginRoute } from '../components/Auth/LoginRoute';
import { Layout } from './UI';

export class App extends Component {
  componentDidMount = () => {
    console.log(implicitAuthManager.config);
    implicitAuthManager.registerHooks({
      onAuthenticateSuccess: () => this.props.login(),
      onAuthenticateFail: () => this.props.logout(),
    });
    // don't call function if on localhost
    if (!window.location.host.match(/localhost/)) {
      implicitAuthManager.handleOnPageLoad();
    }
  };

  render() {
    return (
      <Layout authentication={this.props.authentication}>
        <Switch>
          <Route path="/notAuthorized" component={Restricted} />
          <Route path="/login/:idp" component={LoginRoute} />
          <Route path="/Request/new" component={NewRequest} />
          {/* TODO: use it as /request/:id to make a unique page for redirecting ? */}
          {/* <Route path="/Request/:id" component={Request} /> */}
          <Route path="/home" component={Home} />
          <Route path="/" component={Auth} />
        </Switch>
      </Layout>
    );
  }
}

const mapStateToProps = state => {
  return {
    authentication: state.authentication,
    getRequests: state.getRequests,
    authorization: state.authorization,
    newRequest: state.newRequest,
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      login: () => dispatch(authenticateSuccess()),
      logout: () => dispatch(authenticateFailed()),
    },
    dispatch
  );
};

App.propTypes = {
  authentication: PropTypes.object.isRequired,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
