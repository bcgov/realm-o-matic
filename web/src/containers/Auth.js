import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { ACCESS_CONTROL } from '../constants/auth';
import { SpinLoader } from '../components/UI';
import { AuthModal } from '../components/Auth/AuthModal';
import { authorizationAction } from '../actionCreators';

export class Auth extends Component {
  static displayName = '[Component Auth]';
  componentDidMount = () => {
    const {
      isAuthenticated,
      userInfo,
      userId,
      authorizationStarted,
      authCode,
      errorMessage,
      authorizationAction,
    } = this.props;
    if (isAuthenticated && !authorizationStarted && !authCode && !errorMessage) {
      authorizationAction(userId, userInfo.roles);
    }
  };
  render() {
    const { isAuthenticated, authorizationStarted, authCode, errorMessage } = this.props;

    let redirect = null;
    if (authCode === ACCESS_CONTROL.NO_ROLE) redirect = <Redirect to="/notAuthorized" />;
    else if (
      authCode === ACCESS_CONTROL.REQUESTER_ROLE ||
      authCode === ACCESS_CONTROL.REVIEWER_ROLE
    ) {
      redirect = <Redirect to="/home" />;
    }

    let content;
    if (!isAuthenticated) {
      content = <AuthModal isAuthenticated={isAuthenticated} />;
    } else {
      if (authorizationStarted) content = <SpinLoader text="Loading your user info..." />;
      else if (errorMessage) content = <p>{errorMessage}</p>;
      else content = redirect;
    }
    return <div>{content}</div>;
  }
}

const mapStateToProps = state => {
  return {
    // Authentication:
    isAuthenticated: state.authentication.isAuthenticated,
    userInfo: state.authentication.userInfo,
    userId: state.authentication.userId,
    // Authorization:
    authorizationStarted: state.authorization.authorizationStarted,
    authCode: state.authorization.authCode,
    errorMessage: state.authorization.errorMessage,
  };
};

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      authorizationAction,
    },
    dispatch
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Auth);
