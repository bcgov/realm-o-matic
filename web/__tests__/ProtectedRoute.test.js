import React, { Component } from 'react';
import { render, cleanup } from 'react-testing-library';
import { BrowserRouter } from 'react-router-dom';
import { ProtectedRoute } from '../src/components/Auth';
import { ACCESS_CONTROL } from '../src/constants/auth';

// jest.mock('../src/containers/Auth.js', () => <h1>123</h1>);

afterEach(cleanup);

describe('ProtectedRoute Component', () => {
  class TestComponent extends Component {
    render() {
      return <h1>test content</h1>;
    }
  }

  it('render content when user has correct role', () => {
    const { container } = render(
      <BrowserRouter>
        <ProtectedRoute component={TestComponent} authCode={ACCESS_CONTROL.REQUESTER_ROLE} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  // Issue with testing on redirec: Maximum update depth exceeded.
  it.skip('redirects without correct role', () => {
    const { container } = render(
      <BrowserRouter>
        <ProtectedRoute component={TestComponent} authCode={ACCESS_CONTROL.NO_ROLE} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
