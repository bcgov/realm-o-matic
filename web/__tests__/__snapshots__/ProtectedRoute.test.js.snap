// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProtectedRoute Component render content when user has correct role 1`] = `
<div>
  <h1>
    test content
  </h1>
</div>
`;
